lib/logstash/pluginmanager/install.rb

```diff
@@ -18,95 +19,119 @@ class LogStash::PluginManager::Install < Clamp::Command
-  #
-  # TODO: find right syntax to allow specifying list of plugins with optional version specification for each
-
   def execute
-    if development?
-      raise(LogStash::PluginManager::Error, "Cannot specify plugin(s) with --development, it will add the development dependencies of the currently installed plugins") unless plugin_list.empty?
+    validate_cli_options!
+    
+    if local_gems?
+      gems = extract_local_gems_plugins
+    elsif development?
+      gems = plugins_development_gems
     else
-      raise(LogStash::PluginManager::Error, "No plugin specified") if plugin_list.empty? && verify?
-
-      # temporary until we fullfil TODO ^^
-      raise(LogStash::PluginManager::Error, "Only 1 plugin name can be specified with --version") if version && plugin_list.size > 1
+      gems = plugins_gems
+      verify!(gems)
     end
-    raise(LogStash::PluginManager::Error, "File #{LogStash::Environment::GEMFILE_PATH} does not exist or is not writable, aborting") unless File.writable?(LogStash::Environment::GEMFILE_PATH)
-
-    gemfile = LogStash::Gemfile.new(File.new(LogStash::Environment::GEMFILE_PATH, "r+")).load
-    # keep a copy of the gemset to revert on error
-    original_gemset = gemfile.gemset.copy
 
-    # force Rubygems sources to our Gemfile sources
-    Gem.sources = gemfile.gemset.sources
-
-    # install_list will be an array of [plugin name, version] tuples, version can be nil
-    install_list = []
+    install_gems_list!(gems)
+    remove_unused_locally_installed_gems! 
+  end
 
+  private
+  def validate_cli_options!
     if development?
-      specs = LogStash::PluginManager.all_installed_plugins_gem_specs(gemfile)
-      install_list = specs.inject([]) do |result, spec|
-        result = result + spec.dependencies.select{|dep| dep.type == :development}.map{|dep| [dep.name] + dep.requirement.as_list + [{:group => :development}]}
-      end
+      signal_usage_error("Cannot specify plugin(s) with --development, it will add the development dependencies of the currently installed plugins") unless plugin_list.empty?
     else
-      # at this point we know that plugin_list is not empty and if the --version is specified there is only one plugin in plugin_list
-
-      install_list = version ? [plugin_list << version] : plugin_list.map{|plugin| [plugin, nil]}
+      signal_usage_error("No plugin specified") if plugin_list.empty? && verify?
+      # TODO: find right syntax to allow specifying list of plugins with optional version specification for each
+      signal_usage_error("Only 1 plugin name can be specified with --version") if version && plugin_list.size > 1
+    end
+    signal_error("File #{LogStash::Environment::GEMFILE_PATH} does not exist or is not writable, aborting") unless ::File.writable?(LogStash::Environment::GEMFILE_PATH)
+  end
 
-      install_list.each do |plugin, version|
+  # Check if the specified gems contains
+  # the logstash `metadata`
+  def verify!(gems)
```

Comment:
I meant, since it stop the flow of exception, I think it is a good candidate for the `bang!`